Let's import all libraries I would need for cleaning the data.

In [55]:
import numpy as np
import pandas as pd
import glob
import os

I have data divided in four files based on victim count. Let's see the first five rows for exemplary file.

In [56]:
df_highest = pd.read_csv('./Highest_victim_count.csv')
df_highest.head()

Name                                  Country  \
0           Luis Garavito         Colombia\r\nEcuador\r\nVenezuela   
1             Pedro López              Colombia\r\nPeru\r\nEcuador   
2             Javed Iqbal                                 Pakistan   
3          Mikhail Popkov                                   Russia   
4  Daniel Camargo Barbosa  Colombia\r\nEcuador\r\nBrazil (alleged)   

   Years active  Proven victims Possible victims  \
0  1992 to 1999             138         172–300+   
1  1969 to 1979             110             300+   
2  1998 to 1999             100              100   
3  1992 to 2010              83              83+   
4  1974 to 1986              72              180   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             Notes  
0                                                           Child-murderer, torture-killer, and rapist known as "La Bestia" ("The Beast"). Confessed to killing 140 boys between eight and 16 years old over a seven-year period in Colombia and neighboring countries. He is suspected of murdering over 300 victims, mostly street children. Garavito was originally sentenced to 1,853 years in prison, but this was later reduced to 22 years after he led police to many of the bodies of his victims. He is scheduled to become eligible for parole in 2023.  
1                                                                                                                                                                                              Child-murderer and rapist, known as "The Monster of the Andes". Targeted young girls, between the ages of eight and 12. Arrested in 1980 and convicted in 1983 of killing three young girls, but claimed to have killed hundreds. Despite being believed to be one of the most prolific serial killers of the twentieth century, he was released in the late 1990s.  
2  Child-murderer and rapist, known as "Kukri", Iqbal murdered 100 street children by strangling them and covering up his crimes by dissolving the bodies with acid. He was arrested in 1999 thanks to a letter he sent to a newspaper, and was set to be executed in the manner described by the judge who stated, "You will be strangled to death in front of the parents whose children you killed, Your body will then be cut into 100 pieces and put in acid, the same way you killed the children." However, he died in custody before he could be executed.  
3                                                                                                                                                                                      Serial rapist-killer nicknamed "The Werewolf", who was active for two decades in Angarsk, Irkutsk and Vladivostok. After being convicted of 22 murders in 2015, he confessed to an additional 59 murders, of which he was convicted of 56 in 2018. In July 2020, Popkov confessed to more killings, bringing the total number of admitted victims to 83. Sentenced to life.  
4                                                                Child and woman murderer, believed to have possibly raped and killed over 150 victims, primarily targeting female children as they were more likely to be virgins. Confessed to killing 72 victims. He strangled young girls in Colombia and was arrested, but he escaped from prison and he started killing in Ecuador; rearrested in 1986, he was incarcerated in the same Ecuadorian prison as 300+ serial killer Pedro López. Camargo was killed in jail by the nephew of one of his vict

I can clearly see that at least "Country" and "Possible victims" column will need my intervention.

However, I can't see the whole "Notes" column. I will fix it with few display functions seen below so I can get a full picture of a dataset.

In [57]:
pd.options.display.max_columns = None
pd.options.display.max_rows = None
pd.options.display.max_colwidth = None

In [58]:
df_highest.head()

Name                                  Country  \
0           Luis Garavito         Colombia\r\nEcuador\r\nVenezuela   
1             Pedro López              Colombia\r\nPeru\r\nEcuador   
2             Javed Iqbal                                 Pakistan   
3          Mikhail Popkov                                   Russia   
4  Daniel Camargo Barbosa  Colombia\r\nEcuador\r\nBrazil (alleged)   

   Years active  Proven victims Possible victims  \
0  1992 to 1999             138         172–300+   
1  1969 to 1979             110             300+   
2  1998 to 1999             100              100   
3  1992 to 2010              83              83+   
4  1974 to 1986              72              180   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             Notes  
0                                                           Child-murderer, torture-killer, and rapist known as "La Bestia" ("The Beast"). Confessed to killing 140 boys between eight and 16 years old over a seven-year period in Colombia and neighboring countries. He is suspected of murdering over 300 victims, mostly street children. Garavito was originally sentenced to 1,853 years in prison, but this was later reduced to 22 years after he led police to many of the bodies of his victims. He is scheduled to become eligible for parole in 2023.  
1                                                                                                                                                                                              Child-murderer and rapist, known as "The Monster of the Andes". Targeted young girls, between the ages of eight and 12. Arrested in 1980 and convicted in 1983 of killing three young girls, but claimed to have killed hundreds. Despite being believed to be one of the most prolific serial killers of the twentieth century, he was released in the late 1990s.  
2  Child-murderer and rapist, known as "Kukri", Iqbal murdered 100 street children by strangling them and covering up his crimes by dissolving the bodies with acid. He was arrested in 1999 thanks to a letter he sent to a newspaper, and was set to be executed in the manner described by the judge who stated, "You will be strangled to death in front of the parents whose children you killed, Your body will then be cut into 100 pieces and put in acid, the same way you killed the children." However, he died in custody before he could be executed.  
3                                                                                                                                                                                      Serial rapist-killer nicknamed "The Werewolf", who was active for two decades in Angarsk, Irkutsk and Vladivostok. After being convicted of 22 murders in 2015, he confessed to an additional 59 murders, of which he was convicted of 56 in 2018. In July 2020, Popkov confessed to more killings, bringing the total number of admitted victims to 83. Sentenced to life.  
4                                                                Child and woman murderer, believed to have possibly raped and killed over 150 victims, primarily targeting female children as they were more likely to be virgins. Confessed to killing 72 victims. He strangled young girls in Colombia and was arrested, but he escaped from prison and he started killing in Ecuador; rearrested in 1986, he was incarcerated in the same Ecuadorian prison as 300+ serial killer Pedro López. Camargo was killed in jail by the nephew of one of his vict

Like it was said before, the dataset contains four files divided by victim count. 

I want to join them into one file using glob and os (it will find all files that matches the concrete pathname - in this case "victim_count") and pd.concat function.

In [59]:
data_files = glob.glob(os.path.join('', './*victim_count.csv'))
df = pd.concat(map(lambda file: pd.read_csv(file), data_files),ignore_index=True)

In [60]:
df.isnull().sum()

Name                 0
Country              0
Years active         0
Proven victims       0
Possible victims    68
Notes                0
dtype: int64

In [61]:
df.shape

(305, 6)

I've got 305 rows and 6 columns in our dataframe and I have missing values (68) in only one column ("Possible victims") - so far, so good. 

First and foremost I would like to clear out the "Proven victims" and "Possible victims" column.

It came to my attention almost immediately that these two columns can provide me with some significant insights and interesting analysis results. 

And from the first look at the "Possible victims" data I can tell that there is a lot of work to do.

In [62]:
df['Possible victims'].value_counts(dropna=False)

NaN         68
16          11
10           9
11           9
15           9
17           8
19           7
20           7
12           5
10+          5
Unknown      5
25           5
22           5
19+          5
13           5
21           4
18           4
40+          4
20+          4
18+          3
13+          3
100          3
30           3
14           3
14+          3
25+          3
50+          3
48           2
29           2
33           2
-            2
15+          2
11+          2
100+         2
16+          2
36+          2
27+          2
26           2
3            2
4            2
23+          2
43+          2
23           2
4+           2
30+          2
44           2
35+          2
60+          1
52+          1
50           1
67           1
80-300       1
93           1
49           1
71–90+       1
9-100+       1
70+          1
56           1
60           1
8+           1
~39          1
70–80+       1
41           1
9–25         1
32+          1
76           1
80–100    

Above function let me group the values and show them in a more see-through way.

I can see that I have a lot of missing values ('NaN') in the dataset in this particular column - 68 to be exact. 

I have also values with different characters added to integers like '+', '?' or "~". 

I can see few spots where the value is described as "Unknown" or "-".

There is also quite a lot of values in intervals. 

The "Possible victims" column even from a name is indicating that the data in it is speculative. 

I will make a few assumptions and clean the data in accordance to them:

- the 'Nan' values will be replaced with corresponding values from the "Proven victims" column,
- the "Unknown" and "-" values will be replaced with corresponding values from the "Proven victims" column, 
- when it comes to values in intervals, I will leave the highest number from the interval.
- the values with '+', '?' or "~", I will erase the special characters and assume the integer as a correct value,


Let's localize missing values in "Possible victims" and display them alongside the "Proven victims" column - I already know that the missing values only occur in the "Possible victims" column.

In [63]:
df.loc[df['Possible victims'].isna(), ['Proven victims','Possible victims']] 

Proven victims Possible victims
66              15              NaN
67              15              NaN
82              14              NaN
83              14              NaN
84              14              NaN
85              14              NaN
86              14              NaN
87              14              NaN
88              14              NaN
89              14              NaN
90              14              NaN
100             13              NaN
101             13              NaN
102             13              NaN
103             13              NaN
104             13              NaN
105             13              NaN
106             13              NaN
107             13              NaN
108             13              NaN
109             13              NaN
110             13              NaN
111             13              NaN
112             13              NaN
113             13              NaN
122             12              NaN
123             12              NaN
124             12              NaN
125             12              NaN
126             12              NaN
129             11              NaN
135             11              NaN
136             11              NaN
137             11              NaN
139             11              NaN
141             11              NaN
142             11              NaN
143             11              NaN
144             11              NaN
145             11              NaN
146             11              NaN
147             11              NaN
148             11              NaN
149             11              NaN
150             11              NaN
151             11              NaN
152             11              NaN
162             10              NaN
163             10              NaN
164             10              NaN
165             10              NaN
166             10              NaN
167             10              NaN
168             10              NaN
169             10              NaN
170             10              NaN
171             10              NaN
172             10              NaN
173             10              NaN
174             10              NaN
190              9              NaN
203              8              NaN
204              8              NaN
218              7              NaN
219              7              NaN
220              6              NaN
221              6              NaN
239              5              NaN

Now, I will fill up the missing values with the values from corresponding rows in "Proven victims" column.

In [64]:
df['Possible victims'] = df['Possible victims'].fillna(df['Proven victims'])

In [65]:
df.isnull().sum()

Name                0
Country             0
Years active        0
Proven victims      0
Possible victims    0
Notes               0
dtype: int64

It worked - there is no missing values in the "Possible victims" column. 

Now, I will go right into my next task - replacing the "Unknown" and "-" values with corresponding values from the "Proven victims" column.

Let's localize these values in "Possible victims" and display them alongside the "Proven victims" column.

In [66]:
df.loc[df['Possible victims'].isin(['Unknown', '-']), ['Name','Proven victims','Possible victims']] 

Name Proven victims Possible victims
127      Enriqueta Martí             12          Unknown
138  Henri Désiré Landru             11          Unknown
140       Anthony Sowell             11          Unknown
192     Eric Edgar Cooke              8          Unknown
194       Yoshio Kodaira          8–11+          Unknown
241          Sek Kim Wah              5                -
300  Leonarda Cianciulli              3                -

Next step is filling up the values with "Unknown" and "-" from "Possible victims" with the values from corresponding rows from the "Proven victims" column.

In [67]:
df.loc[(df['Possible victims'].isin(['Unknown', '-']))&(df['Proven victims'].notnull()), 'Possible victims'] = df['Proven victims']

Let's check if the filling up was carried out correctly for one individual - for example Yoshio Kodaira. 

In [68]:
df[df['Name'].str.contains('Yoshio Kodaira')]

Name         Country   Years active Proven victims  \
194  Yoshio Kodaira  China\r\nJapan  1928? to 1946          8–11+   

    Possible victims  \
194            8–11+   

                                                                                                                                                                                                                                                                                                                            Notes  
194  A serial rapist, Kodaira killed his father-in-law in 1932 and eight to 10 women in Japan between 1945 and 1946, engaging in necrophilia after the fifth murder. Previously (1920s) he had been deployed to Northern China as a sailor in the Imperial Japanese Navy, where he was free to target the locals. Hanged in 1949.

I can see that the value from the "Proven victims" column was copied into "Possible victims" column for that row. 

Now, I will clean up the values in intervals. 

From what I saw intervals consist of integer, dash sign and other integer, let's display these lines.

In [69]:
df[df['Possible victims'].str.contains('-') == True]

Name        Country  Years active Proven victims  \
251  Giuseppe Greco          Italy  1981 to 1983             58   
302     Albert Fish  United States  1924 to 1932              3   

    Possible victims  \
251           80-300   
302           9-100+   

                                                                                                                                                                                                                                                                                                Notes  
251  During the Second Mafia War, from 1981 to 1983 committed dozens of murders, often with his favorite weapon, an AK-47. He was eventually convicted in absentia of 58 murders, but it is believed that he committed at least 80 murders and possibly as many as 300. Murdered in his home in 1985.  
302                                                                                                                                                                         American serial killer and child rapist who confessed to three murders for which he was convicted and sentenced to death.

There was definitely a lot more values in intervals when I carried out 'values.count' function. 

I went back to check it and it seems that there is a longer dash used in few of the intervals.

Let's display these lines.

In [70]:
df[df['Possible victims'].str.contains('–') == True]

Name                           Country  \
51      Randy Steven Kraft                     United States   
115   Donald Henry Gaskins                     United States   
119          Dennis Nilsen                    United Kingdom   
131          Juana Barraza                            Mexico   
132         Zhang Yongming                             China   
194         Yoshio Kodaira                    China\r\nJapan   
205   Nikolai Dzhumagaliev                      Soviet Union   
208             Ivan Milat                         Australia   
232     Richard Cottingham                     United States   
234  John Floyd Thomas Jr.                     United States   
236           Steve Wright                    United Kingdom   
242          Luis Garavito  Colombia\r\nEcuador\r\nVenezuela   
257           Gary Ridgway                     United States   
260          Ahmad Suradji                         Indonesia   
264           Serhiy Tkach           Soviet Union\r\nUkraine   
265    Gennady Mikhasevich                      Soviet Union   
268              Ted Bundy                     United States   
298        Dorothea Puente                     United States   

           Years active Proven victims Possible victims  \
51         1971 to 1983             16            65–67   
115        1953 to 1982             12           31–80+   
119        1978 to 1983             12            15–16   
131  late 1990s to 2006             11            29–49   
132        2008 to 2012             11            17–20   
194       1928? to 1946          8–11+            8–11+   
205    1980 and earlier              7           50–100   
208               1990s              7            23–37   
232        1977 to 1980              5           85–100   
234        1957 to 2009              5            17–25   
236                2006              5             5–22   
242        1992 to 1999            138         172–300+   
257        1982 to 2000             49           71–90+   
260        1986 to 1997             42           70–80+   
264        1984 to 2005             36           80–100   
265        1971 to 1985             36           43–55+   
268        1974 to 1978             35          36–100+   
298        1982 to 1988              3             9–25   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Notes  
51                                                                                                                                                                                                                                                                                                                                                       Convicted of 16 counts of murder but left a cryptic "score card" referring to at least 65 victims. May have had an accomplice. Sentenced to death.  
115                                                                                                              Known as "The Meanest Man in America", Gaskins was convicted of nine murders committed in South Carolina between 1973 and 1975. He was suspected of 31 murders. Two victims had been murdered while Gaskins had been incarcerated—one while Gaskins had been on death row. Later claimed on death row that he had murdered between 80 and 110 victims. Executed by electric chair in 1991.  
119                                                                                                                                                                                                      

It seems that now I have all the lines I need. 

Now, let's do the split for each sign so I can have only the highest number from interval in the "Possible victims" column. 

In [71]:
df['Possible victims'] = [str(x).split('-')[-1] for x in df['Possible victims']]
df['Possible victims'] = [str(x).split('–')[-1] for x in df['Possible victims']]

I will check if I am left with any values in intervals for "Possible victims" column.

In [72]:
df['Possible victims'].value_counts(dropna=False)

11      28
10      22
13      19
14      12
16      12
12      11
15      11
20       8
17       8
25       7
19       7
100      6
22       6
19+      5
10+      5
21       4
20+      4
100+     4
40+      4
18       4
11+      3
50+      3
14+      3
80+      3
8        3
25+      3
13+      3
3        3
30       3
7        3
18+      3
6        2
300+     2
5        2
48       2
37       2
29       2
49       2
15+      2
36+      2
4        2
26       2
23+      2
23       2
43+      2
27+      2
35+      2
33       2
30+      2
44       2
67       2
16+      2
4+       2
52+      1
32       1
50       1
90+      1
56       1
8+       1
~39      1
60       1
60+      1
41       1
34+      1
76       1
55+      1
32+      1
~35      1
35       1
93       1
300      1
42+      1
70+      1
180      1
45       1
28       1
24+      1
24       1
150      1
26+      1
82+      1
21+      1
38+      1
42       1
79       1
34       1
9        1
160      1
130+     1
12+      1
70?      1

I am not! That's great news. 

The only thing that I am left to clean in this particular column are the special characters '+', '?' or "~". 

In [73]:
df["Possible victims"] = df["Possible victims"].str.replace("+","")
df["Possible victims"] = df["Possible victims"].str.replace("?","")
df["Possible victims"] = df["Possible victims"].str.replace("~","")

C:\Users\Gaba Mach\AppData\Local\Temp\ipykernel_65428\841394383.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df["Possible victims"] = df["Possible victims"].str.replace("+","")
C:\Users\Gaba Mach\AppData\Local\Temp\ipykernel_65428\841394383.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df["Possible victims"] = df["Possible victims"].str.replace("?","")


Let's check if all the values in the "Possible victims" column are cleaned up.

In [74]:
df['Possible victims'].value_counts(dropna=False)

11     31
10     27
13     22
14     15
16     14
15     13
20     12
12     12
19     12
100    10
25     10
17      8
18      7
22      6
30      5
21      5
40      4
8       4
4       4
3       4
50      4
23      4
35      4
80      3
300     3
6       3
26      3
7       3
44      2
34      2
32      2
33      2
60      2
5       2
70      2
37      2
48      2
42      2
36      2
67      2
49      2
27      2
24      2
29      2
43      2
52      1
45      1
55      1
76      1
39      1
41      1
90      1
79      1
56      1
93      1
9       1
83      1
38      1
150     1
82      1
130     1
160     1
28      1
180     1
Name: Possible victims, dtype: int64

It's all cleaned up! 

I have managed to do that without erasing any row of the data. 

I assumed it would be a good choice due to a speculative nature of the data in this particular column. 

Now, let's check the "Proven victims" column.

In [75]:
df['Proven victims'].value_counts(dropna=False)

11       24
13       23
10       22
3        18
9        16
16       15
7        15
12       14
14       13
15       12
6        12
8        11
4        11
19       10
5        10
18        7
21        6
17        6
20        4
33        4
35        3
36        3
30        3
24        3
23        3
32        2
25        2
50        2
49        2
28        2
27        2
22        2
11–16     1
38        1
39        1
41        1
42        1
45        1
48        1
26        1
53        1
52        1
8–15      1
58        1
60        1
67        1
70        1
71        1
83        1
100       1
110       1
138       1
9–13      1
8–11+     1
72        1
Name: Proven victims, dtype: int64

Right of the bat, I can see that the data is way cleaner that in the "Possible victims" column.

However, I still have some values in intervals and one "+" character beside integer despite the fact that the data in this column should be a confirmed and concrete one. 

It may be due to many things - inconsistencies in the sources, lack of or incomplete documentation of the trials for each murder etc. 

I can either erase the rows with such values or include them.

I decided to include them with below assumptions: 

- when it comes to values in intervals, I will leave the highest number from the interval,
- the values with '+' - I will erase the special character and assume the integer as a correct value.

Let's check the values in intervals.

I will check both signs - small dash and longer dash - like in "Possible victims" column.

In [76]:
df[df['Proven victims'].str.contains('-') == True]

Empty DataFrame
Columns: [Name, Country, Years active, Proven victims, Possible victims, Notes]
Index: []

In [77]:
df[df['Proven victims'].str.contains('–') == True]

Name  \
128       Herb Baumeister   
175         Volker Eckert   
193  Joseph Paul Franklin   
194        Yoshio Kodaira   

                                                                                           Country  \
128                                                                                  United States   
175  East Germany\r\nGermany\r\nFrance\r\nSpain\r\nCzech Republic (suspected)\r\nItaly (suspected)   
193                                                                                  United States   
194                                                                                 China\r\nJapan   

      Years active Proven victims Possible victims  \
128   1990 to 1996          11–16               25   
175   1974 to 2006           9–13               19   
193   1977 to 1980           8–15               20   
194  1928? to 1946          8–11+               11   

                                                                                                                                                                                                                                                                                                                                                        Notes  
128                                                                                                                                                 Strangled gay men and buried their bodies in his backyard in Indiana and Ohio; 11 men were found in the yard but only five were identified. Committed suicide by shooting himself when faced with arrest.  
175  German trucker who confessed to having abducted, tortured and killed five prostitutes through his route in Western Europe, plus strangling a 14-year-old girl in his native West Germany in 1974, when he was 15. Police considered him perpetrator of nine murders (with four more being possible); he hanged himself in prison before being convicted.  
193                                                                                                                                                                                                                                White supremacist shooter who confessed to 20 murders and several attempted murders. Executed by lethal injection in 2013.  
194                              A serial rapist, Kodaira killed his father-in-law in 1932 and eight to 10 women in Japan between 1945 and 1946, engaging in necrophilia after the fifth murder. Previously (1920s) he had been deployed to Northern China as a sailor in the Imperial Japanese Navy, where he was free to target the locals. Hanged in 1949.

It seems that there are only intervals with longer dash. 

Let's split the values up and leave the highest one from interval.

In [78]:
df['Proven victims'] = [str(x).split('–')[-1] for x in df['Proven victims']]

I will erase the "+" sign now. 

In [79]:
df["Proven victims"] = df["Proven victims"].str.replace("+","")

C:\Users\Gaba Mach\AppData\Local\Temp\ipykernel_65428\324378986.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df["Proven victims"] = df["Proven victims"].str.replace("+","")


I will check if there is anything else to clean in this column.

In [80]:
df['Proven victims'].value_counts(dropna=False)

11     25
13     24
10     22
3      18
9      16
16     16
7      15
12     14
14     13
15     13
6      12
8      11
4      11
19     10
5      10
18      7
17      6
21      6
20      4
33      4
36      3
35      3
30      3
23      3
24      3
22      2
27      2
32      2
25      2
49      2
50      2
28      2
70      1
67      1
110     1
26      1
100     1
83      1
38      1
39      1
41      1
42      1
45      1
48      1
72      1
71      1
52      1
53      1
58      1
60      1
138     1
Name: Proven victims, dtype: int64

It seems not! The "Proven victims" and "Possible victims" column are all good and clean. 